backupccl: `TestTenantBackupNemesis` very flaky

On my machine this is trivially reproducible with `./dev test pkg/ccl/backupccl -f TestTenantBackupNemesis --stress`:

```
panic: /Tenant/10/Table/3/1/110/2/1: failed to unmarshal descriptor {[0 0 0 4 101 10 0 32 1 51 119 247 59 10 35 0 22 100 66 77 65 76 118 75 78 100 78 121 105 66 116 83 67 113 87 79 69 66 66 72 120 89 108 85 70 122 74 67 75 102 101 121 65 81 65 120 102 106 75 79 104 101 99 105 107 66 83 77 107 90 72 82 106 68 75 67 116 66 102 97 69 83 68 73 121 115 68 107 77 108 116 114 77 122 108 90 121 102 110 72 119 107 111 98 99 108 87 108 76 100 77 83 66 87 112 84 97 72 1 16 64 24 0 48 0 80 20 96 0 32 1 48 0 104 0 112 0 120 0 128 1 0 136 1 0 152 1 0 66 38 10 1 115 16 3 26 12 8 7 16 0 24 0 48 0 80 25 96 0 32 1 48 0 104 0 112 0 120 0 128 1 0 136 1 0 152 1 0 72 4 82 164 1 10 54 99 114 101 97 116 101 95 105 110 100 101 120 95 56 97 53 97 51 56 53 101 45 98 48 98 56 45 52 57 55 55 45 56 51 101 49 45 53 49 57 51 53 53 98 99 97 51 56 98 95 112 107 101 121 16 1 24 1 34 2 105 100 42 1 110 42 1 115 48 1 64 0 74 16 8 0 16 0 26 0 32 0 40 0 48 0 56 0 64 0 90 0 112 2 112 3 122 4 8 0 32 0 128 1 0 136 1 0 144 1 4 152 1 1 162 1 6 8 0 18 0 24 0 168 1 0 178 1 0 186 1 0 192 1 0 200 1 144 247 153 157 153 135 156 222 23 208 1 1 224 1 0 233 1 0 0 0 0 0 0 0 0 96 4 106 33 10 11 10 5 97 100 109 105 110 16 2 24 2 10 10 10 4 114 111 111 116 16 2 24 2 18 4 114 111 111 116 24 3 114 167 1 24 3 32 1 40 1 56 0 18 156 1 10 55 99 114 101 97 116 101 95 105 110 100 101 120 95 56 97 53 97 51 56 53 101 45 98 48 98 56 45 52 57 55 55 45 56 51 101 49 45 53 49 57 51 53 53 98 99 97 51 56 98 95 110 95 105 100 120 16 2 24 0 34 1 110 48 2 56 1 64 0 74 16 8 0 16 0 26 0 32 0 40 0 48 0 56 0 64 0 90 0 122 4 8 0 32 0 128 1 0 136 1 1 144 1 4 152 1 0 162 1 6 8 0 18 0 24 0 168 1 0 178 1 0 186 1 0 192 1 0 200 1 176 179 200 194 153 135 156 222 23 208 1 0 224 1 0 233 1 0 0 0 0 0 0 0 0 114 142 1 24 1 32 1 40 1 56 0 18 131 1 10 38 99 114 100 98 95 105 110 116 101 114 110 97 108 95 105 110 100 101 120 95 51 95 110 97 109 101 95 112 108 97 99 101 104 111 108 100 101 114 16 3 24 0 34 1 110 48 2 56 1 64 0 74 16 8 0 16 0 26 0 32 0 40 0 48 0 56 0 64 0 90 0 122 4 8 0 32 0 128 1 0 136 1 1 144 1 4 152 1 0 162 1 6 8 0 18 0 24 0 168 1 0 178 1 0 186 1 0 192 1 1 200 1 0 208 1 1 224 1 0 233 1 0 0 0 0 0 0 0 0 128 1 1 136 1 3 152 1 0 178 1 29 10 7 112 114 105 109 97 114 121 16 0 26 2 105 100 26 1 110 26 1 115 32 1 32 2 32 3 40 0 184 1 1 194 1 0 232 1 0 242 1 4 8 0 18 0 248 1 0 128 2 0 146 2 0 154 2 10 8 200 142 154 157 153 135 156 222 23 178 2 0 184 2 0 192 2 105 200 2 0 224 2 0 128 3 0 136 3 2 146 3 157 6 10 13 10 7 234 2 4 8 110 16 1 18 0 24 2 10 13 10 7 154 8 4 8 110 16 104 18 0 24 2 10 25 10 15 186 1 12 10 10 8 110 16 2 168 1 1 176 1 3 18 4 8 1 16 1 24 2 10 19 10 9 226 2 6 8 110 16 2 24 2 18 4 8 1 16 1 24 2 10 21 10 11 226 2 8 8 110 16 2 24 1 40 1 18 4 8 1 16 1 24 2 10 17 10 7 234 2 4 8 110 16 2 18 4 8 1 16 1 24 2 10 74 10 64 194 2 61 8 110 16 2 26 55 99 114 101 97 116 101 95 105 110 100 101 120 95 56 97 53 97 51 56 53 101 45 98 48 98 56 45 52 57 55 55 45 56 51 101 49 45 53 49 57 51 53 53 98 99 97 51 56 98 95 110 95 105 100 120 18 4 8 1 16 1 24 2 10 26 10 16 194 1 13 10 9 8 110 16 3 112 1 168 1 1 16 1 18 4 8 1 16 1 24 3 10 19 10 9 226 2 6 8 110 16 3 24 2 18 4 8 1 16 1 24 3 10 21 10 11 226 2 8 8 110 16 3 24 1 40 1 18 4 8 1 16 1 24 3 10 17 10 7 234 2 4 8 110 16 3 18 4 8 1 16 1 24 3 18 193 1 10 190 1 10 71 67 82 69 65 84 69 32 73 78 68 69 88 32 79 78 32 34 99 114 101 97 116 101 95 105 110 100 101 120 95 56 97 53 97 51 56 53 101 45 98 48 98 56 45 52 57 55 55 45 56 51 101 49 45 53 49 57 51 53 53 98 99 97 51 56 98 34 32 40 110 41 18 101 67 82 69 65 84 69 32 73 78 68 69 88 32 79 78 32 226 128 185 98 97 110 107 226 128 186 46 112 117 98 108 105 99 46 226 128 185 34 99 114 101 97 116 101 95 105 110 100 101 120 95 56 97 53 97 51 56 53 101 45 98 48 98 56 45 52 57 55 55 45 56 51 101 49 45 53 49 57 51 53 53 98 99 97 51 56 98 34 226 128 186 32 40 226 128 185 110 226 128 186 41 26 12 67 82 69 65 84 69 32 73 78 68 69 88 26 6 10 4 114 111 111 116 34 11 2 2 12 2 2 2 2 7 2 2 1 40 129 128 242 211 244 203 222 153 13 50 11 0 1 2 3 4 5 6 7 8 9 10 56 1 74 137 2 8 110 18 49 99 114 101 97 116 101 95 105 110 100 101 120 95 56 97 53 97 51 56 53 101 45 98 48 98 56 45 52 57 55 55 45 56 51 101 49 45 53 49 57 51 53 53 98 99 97 51 56 98 82 6 8 1 18 2 105 100 82 5 8 2 18 1 110 82 5 8 3 18 1 115 82 16 8 254 255 255 255 15 18 8 116 97 98 108 101 111 105 100 82 36 8 255 255 255 255 15 18 28 99 114 100 98 95 105 110 116 101 114 110 97 108 95 109 118 99 99 95 116 105 109 101 115 116 97 109 112 90 11 8 0 18 7 112 114 105 109 97 114 121 98 58 8 1 18 54 99 114 101 97 116 101 95 105 110 100 101 120 95 56 97 53 97 51 56 53 101 45 98 48 98 56 45 52 57 55 55 45 56 51 101 49 45 53 49 57 51 53 53 98 99 97 51 56 98 95 112 107 101 121 98 59 8 2 18 55 99 114 101 97 116 101 95 105 110 100 101 120 95 56 97 53 97 51 56 53 101 45 98 48 98 56 45 52 57 55 55 45 56 51 101 49 45 53 49 57 51 53 53 98 99 97 51 56 98 95 110 95 105 100 120 168 3 0 176 3 0 208 3 0 216 3 1] 1710365351.190232000,1}
```

--- a/pkg/kv/bulk/sst_batcher.go
+++ b/pkg/kv/bulk/sst_batcher.go
@@ -340,7 +340,10 @@ func (b *SSTBatcher) AddMVCCKeyWithImportEpoch(
 		return err
 	}
 	mvccVal.MVCCValueHeader.ImportEpoch = importEpoch
-	b.valueScratch, err = storage.EncodeMVCCValueToBuf(mvccVal, b.valueScratch[:0])
+	buf, canRetainBuffer, err := storage.EncodeMVCCValueToBuf(mvccVal, b.valueScratch[:0])
+	if canRetainBuffer {
+		b.valueScratch = buf
+	}
 	if err != nil {
 		return err
 	}

--- a/pkg/storage/mvcc.go
+++ b/pkg/storage/mvcc.go
@@ -7886,11 +7886,14 @@ func mvccExportToWriter(
 			}
 
 			if !ok && opts.IncludeMVCCValueHeader {
-				valueScratch, err = EncodeMVCCValueForExport(mvccValue, valueScratch[:0])
+				buf, canRetainBuf, err := EncodeMVCCValueForExport(mvccValue, valueScratch[:0])
 				if err != nil {
 					return kvpb.BulkOpSummary{}, ExportRequestResumeInfo{}, errors.Wrapf(err, "repackaging imported mvcc value %s", unsafeKey)
 				}
-				unsafeValue = valueScratch
+				if canRetainBuf {
+					valueScratch = buf
+				}
+				unsafeValue = buf
 			} else {
 				unsafeValue = mvccValue.Value.RawBytes
 			}

--- a/pkg/storage/mvcc_value.go
+++ b/pkg/storage/mvcc_value.go
@@ -137,11 +137,13 @@ func (v MVCCValue) SafeFormat(w redact.SafePrinter, _ rune) {
 // EncodeMVCCValueForExport encodes fields from the MVCCValueHeader
 // that are appropriate for export out of the cluster.
 //
-// NB: Must be updated with ExtendedEncodingSizeForExport.
-func EncodeMVCCValueForExport(mvccValue MVCCValue, b []byte) ([]byte, error) {
+// The returned bool is true if the provided buffer was used or
+// reallocated and false if the MVCCValue.Value.RawBytes were returned
+// directly.
+func EncodeMVCCValueForExport(mvccValue MVCCValue, b []byte) ([]byte, bool, error) {
 	mvccValue.MVCCValueHeader.LocalTimestamp = hlc.ClockTimestamp{}
 	if mvccValue.MVCCValueHeader.IsEmpty() {
-		return mvccValue.Value.RawBytes, nil
+		return mvccValue.Value.RawBytes, false, nil
 	}
 	return EncodeMVCCValueToBuf(mvccValue, b)
 }
@@ -178,7 +180,8 @@ func encodedMVCCValueSize(v MVCCValue) int {
 // EncodeMVCCValue encodes an MVCCValue into its Pebble representation. See the
 // comment on MVCCValue for a description of the encoding scheme.
 func EncodeMVCCValue(v MVCCValue) ([]byte, error) {
-	return EncodeMVCCValueToBuf(v, nil)
+	b, _, err := EncodeMVCCValueToBuf(v, nil)
+	return b, err
 }
 
 // EncodeMVCCValueToBuf encodes an MVCCValue into its Pebble
@@ -189,16 +192,20 @@ func EncodeMVCCValue(v MVCCValue) ([]byte, error) {
 // it is large enough. If the provided buffer is not large enough a
 // new buffer is allocated.
 //
+// The returned bool is true if the provided buffer was used or
+// reallocated and false if the MVCCValue.Value.RawBytes were returned
+// directly.
+//
 // TODO(erikgrinaker): This could mid-stack inline when we compared
 // v.MVCCValueHeader == enginepb.MVCCValueHeader{} instead of IsEmpty(), but
 // struct comparisons have a significant performance regression in Go 1.19 which
 // negates the inlining gain. Reconsider this with Go 1.20. See:
 // https://github.com/cockroachdb/cockroach/issues/88818
-func EncodeMVCCValueToBuf(v MVCCValue, buf []byte) ([]byte, error) {
+func EncodeMVCCValueToBuf(v MVCCValue, buf []byte) ([]byte, bool, error) {
 	if v.MVCCValueHeader.IsEmpty() && !disableSimpleValueEncoding {
 		// Simple encoding. Use the roachpb.Value encoding directly with no
 		// modification. No need to re-allocate or copy.
-		return v.Value.RawBytes, nil
+		return v.Value.RawBytes, false, nil
 	}
 
 	// Extended encoding. Wrap the roachpb.Value encoding with a header containing
@@ -224,11 +231,11 @@ func EncodeMVCCValueToBuf(v MVCCValue, buf []byte) ([]byte, error) {
 	// an interface, which would cause a heap allocation and incur the cost of
 	// dynamic dispatch.
 	if _, err := v.MVCCValueHeader.MarshalToSizedBuffer(buf[extendedPreludeSize:headerSize]); err != nil {
-		return nil, errors.Wrap(err, "marshaling MVCCValueHeader")
+		return nil, false, errors.Wrap(err, "marshaling MVCCValueHeader")
 	}
 	// <4-byte-checksum><1-byte-tag><encoded-data> or empty for tombstone
 	copy(buf[headerSize:], v.Value.RawBytes)
-	return buf, nil
+	return buf, true, nil
 }
 
 // DecodeMVCCValue decodes an MVCCKey from its Pebble representation.

--- a/pkg/storage/mvcc_value_test.go
+++ b/pkg/storage/mvcc_value_test.go
@@ -105,7 +105,7 @@ func TestEncodeMVCCValueForExport(t *testing.T) {
 	}
 	for name, tc := range testcases {
 		t.Run(name, func(t *testing.T) {
-			encodedVal, err := EncodeMVCCValueForExport(tc.val, nil)
+			encodedVal, _, err := EncodeMVCCValueForExport(tc.val, nil)
 			require.NoError(t, err)
 			strippedMVCCVal, err := DecodeMVCCValue(encodedVal)
 			require.NoError(t, err)
@@ -320,7 +320,7 @@ func BenchmarkEncodeMVCCValueForExport(b *testing.B) {
 			b.Run(name, func(b *testing.B) {
 				for i := 0; i < b.N; i++ {
 					var err error
-					buf, err = EncodeMVCCValueForExport(mvccValue, buf[:0])
+					buf, _, err = EncodeMVCCValueForExport(mvccValue, buf[:0])
 					if err != nil { // for performance
 						require.NoError(b, err)
 					}
@@ -342,7 +342,7 @@ func BenchmarkEncodeMVCCValueWithAllocator(b *testing.B) {
 			b.Run(name, func(b *testing.B) {
 				for i := 0; i < b.N; i++ {
 					var err error
-					buf, err = EncodeMVCCValueToBuf(mvccValue, buf[:0])
+					buf, _, err = EncodeMVCCValueToBuf(mvccValue, buf[:0])
 					if err != nil { // for performance
 						require.NoError(b, err)
 					}

